File: packages/engine/src/allergen/backstop.ts
import type {
  CanonicalItem,
  DietaryRestriction,
  Person,
  Recipe,
} from "../schemas/domain.js";
import type { PlanEntryProposal, Violation } from "../schemas/plan.js";
import { itemInAllergenClass, lookupItem } from "./ontology.js";

/**
 * The deterministic allergen backstop.
 *
 * Runs over every proposed plan entry AFTER the LLM, regardless of what the
 * LLM claimed about safety. Policy:
 *  - severity "allergy": any conflicting ingredient is an ERROR unless the
 *    proposed handling provably resolves it (valid substitute for that exact
 *    ingredient, feasible split point, or skip). Unknown ingredients
 *    (unmatched to the ontology) are ALSO errors for allergy severity — we
 *    fail closed.
 *  - severity "intolerance": conflicts must be handled, but unknown
 *    ingredients produce warnings rather than errors.
 *  - severity "preference": conflicts produce warnings only.
 */
export interface BackstopInput {
  entry: PlanEntryProposal;
  entryIndex: number;
  recipe: Recipe;
  people: Person[];
  restrictions: DietaryRestriction[];
  items: ReadonlyMap<string, CanonicalItem>;
}

function ingredientConflicts(
  recipe: Recipe,
  restriction: DietaryRestriction,
  items: ReadonlyMap<string, CanonicalItem>,
): { conflicting: number[]; unknown: number[] } {
  const conflicting: number[] = [];
  const unknown: number[] = [];
  recipe.ingredients.forEach((ing, i) => {
    if (ing.isOptional) return;
    if (restriction.canonicalItemId) {
      if (ing.canonicalItemId === restriction.canonicalItemId) conflicting.push(i);
      else if (ing.canonicalItemId === null) unknown.push(i);
      return;
    }
    if (restriction.allergenClass) {
      const item = lookupItem(items, ing.canonicalItemId);
      const membership = itemInAllergenClass(item, restriction.allergenClass);
      if (membership === "member") conflicting.push(i);
      else if (membership === "unknown") unknown.push(i);
    }
  });
  return { conflicting, unknown };
}

/**
 * A split is feasible only if EVERY conflicting ingredient enters the dish at
 * a known step > 1 — i.e. there is a real moment to pull a clean portion.
 */
export function splitFeasible(recipe: Recipe, conflictingIdx: number[]): boolean {
  if (conflictingIdx.length === 0) return true;
  return conflictingIdx.every((i) => {
    const ing = recipe.ingredients[i];
    return ing !== undefined && ing.addedAtStep !== null && ing.addedAtStep > 1;
  });
}

export function runAllergenBackstop(input: BackstopInput): Violation[] {
  const { entry, entryIndex, recipe, people, restrictions, items } = input;
  const violations: Violation[] = [];

  for (const person of people) {
    if (!person.eatsPlannedDinners) continue;
    const personRestrictions = restrictions.filter((r) => r.personId === person.id);
    for (const restriction of personRestrictions) {
      const { conflicting, unknown } = ingredientConflicts(recipe, restriction, items);
      const handling = entry.personHandling.find((h) => h.personId === person.id);
      const isAllergy = restriction.severity === "allergy";

      // Unknown ingredients: fail closed for allergies, warn otherwise.
      if (unknown.length > 0 && restriction.allergenClass) {
        violations.push({
          code: "allergy_hard_fail",
          severity: isAllergy ? "error" : "warning",
          entryIndex,
          message: `${recipe.title}: ${unknown.length} ingredient(s) not matched to the ontology (${unknown
            .map((i) => recipe.ingredients[i]?.raw)
            .filter(Boolean)
            .join("; ")}) — cannot verify against ${person.name}'s ${restriction.severity} (${restriction.allergenClass}).`,
        });
      }

      if (conflicting.length === 0) continue;

      const conflictNames = conflicting
        .map((i) => recipe.ingredients[i]?.raw)
        .filter(Boolean)
        .join("; ");
      const baseMsg = `${recipe.title}: contains ${conflictNames}, conflicting with ${person.name}'s ${restriction.severity}`;

      if (restriction.severity === "preference") {
        if (!handling || handling.handling === "clear") {
          violations.push({
            code: "allergy_hard_fail",
            severity: "warning",
            entryIndex,
            message: `${baseMsg} (preference — flag to user).`,
          });
        }
        continue;
      }

      // allergy / intolerance: handling must exist and actually resolve it.
      const errSeverity = isAllergy ? ("error" as const) : ("warning" as const);
      if (!handling || handling.handling === "clear") {
        violations.push({
          code: "allergy_hard_fail",
          severity: errSeverity,
          entryIndex,
          message: `${baseMsg}, but handling is missing or "clear".`,
        });
        continue;
      }
      if (handling.handling === "substitute") {
        const sub = handling.substituteItemId
          ? items.get(handling.substituteItemId)
          : undefined;
        const subOk =
          sub !== undefined &&
          (restriction.allergenClass
            ? itemInAllergenClass(sub, restriction.allergenClass) === "non_member"
            : sub.id !== restriction.canonicalItemId);
        if (!subOk) {
          violations.push({
            code: "invalid_substitute",
            severity: errSeverity,
            entryIndex,
            message: `${baseMsg}; proposed substitute is missing, unknown, or itself conflicting.`,
          });
        }
        continue;
      }
      if (handling.handling === "split") {
        if (!splitFeasible(recipe, conflicting)) {
          violations.push({
            code: "split_not_feasible",
            severity: errSeverity,
            entryIndex,
            message: `${baseMsg}; split proposed but a conflicting ingredient enters at step 1 or an unknown step — no clean pull point.`,
          });
        }
        continue;
      }
      // handling === "skip" resolves the conflict for that person.
    }
  }
  return violations;
}

File: service/mealplanner/allergen/backstop.py
"""Deterministic allergen backstop — port of packages/engine/src/allergen/backstop.ts.

Runs over every proposed plan entry AFTER the LLM, regardless of what the LLM
claimed about safety. Policy:
 - severity "allergy": any conflicting ingredient is an ERROR unless the
   proposed handling provably resolves it (valid substitute, feasible split
   point, or skip). Unknown ingredients (unmatched to the ontology) are ALSO
   errors for allergy severity — fail closed.
 - severity "intolerance": conflicts must be handled, but unknown ingredients
   produce warnings rather than errors.
 - severity "preference": conflicts produce warnings only.
"""

from __future__ import annotations

from typing import Mapping

from ..schemas.domain import (
    CanonicalItem,
    DietaryRestriction,
    Person,
    Recipe,
)
from ..schemas.plan import PlanEntryProposal, Violation
from .ontology import item_in_allergen_class, lookup_item


def _ingredient_conflicts(
    recipe: Recipe,
    restriction: DietaryRestriction,
    items: Mapping[str, CanonicalItem],
) -> tuple[list[int], list[int]]:
    conflicting: list[int] = []
    unknown: list[int] = []
    for i, ing in enumerate(recipe.ingredients):
        if ing.is_optional:
            continue
        if restriction.canonical_item_id:
            if ing.canonical_item_id == restriction.canonical_item_id:
                conflicting.append(i)
            elif ing.canonical_item_id is None:
                unknown.append(i)
            continue
        if restriction.allergen_class:
            item = lookup_item(items, ing.canonical_item_id)
            membership = item_in_allergen_class(item, restriction.allergen_class)
            if membership == "member":
                conflicting.append(i)
            elif membership == "unknown":
                unknown.append(i)
    return conflicting, unknown


def split_feasible(recipe: Recipe, conflicting_idx: list[int]) -> bool:
    """A split is feasible only if EVERY conflicting ingredient enters the dish
    at a known step > 1 — i.e. there is a real moment to pull a clean portion."""
    if not conflicting_idx:
        return True
    return all(
        (ing := recipe.ingredients[i]).added_at_step is not None
        and ing.added_at_step > 1
        for i in conflicting_idx
    )


def run_allergen_backstop(
    *,
    entry: PlanEntryProposal,
    entry_index: int,
    recipe: Recipe,
    people: list[Person],
    restrictions: list[DietaryRestriction],
    items: Mapping[str, CanonicalItem],
) -> list[Violation]:
    violations: list[Violation] = []

    for person in people:
        if not person.eats_planned_dinners:
            continue
        for restriction in (r for r in restrictions if r.person_id == person.id):
            conflicting, unknown = _ingredient_conflicts(recipe, restriction, items)
            handling = next(
                (h for h in entry.person_handling if h.person_id == person.id), None
            )
            is_allergy = restriction.severity == "allergy"

            # Unknown ingredients: fail closed for allergies, warn otherwise.
            if unknown and restriction.allergen_class:
                unknown_names = "; ".join(
                    recipe.ingredients[i].raw for i in unknown
                )
                violations.append(
                    Violation(
                        code="allergy_hard_fail",
                        severity="error" if is_allergy else "warning",
                        entry_index=entry_index,
                        message=(
                            f"{recipe.title}: {len(unknown)} ingredient(s) not matched to "
                            f"the ontology ({unknown_names}) — cannot verify against "
                            f"{person.name}'s {restriction.severity} ({restriction.allergen_class})."
                        ),
                    )
                )

            if not conflicting:
                continue

            conflict_names = "; ".join(recipe.ingredients[i].raw for i in conflicting)
            base_msg = (
                f"{recipe.title}: contains {conflict_names}, conflicting with "
                f"{person.name}'s {restriction.severity}"
            )

            if restriction.severity == "preference":
                if handling is None or handling.handling == "clear":
                    violations.append(
                        Violation(
                            code="allergy_hard_fail",
                            severity="warning",
                            entry_index=entry_index,
                            message=f"{base_msg} (preference — flag to user).",
                        )
                    )
                continue

            # allergy / intolerance: handling must exist and actually resolve it.
            err_severity = "error" if is_allergy else "warning"
            if handling is None or handling.handling == "clear":
                violations.append(
                    Violation(
                        code="allergy_hard_fail",
                        severity=err_severity,
                        entry_index=entry_index,
                        message=f'{base_msg}, but handling is missing or "clear".',
                    )
                )
                continue
            if handling.handling == "substitute":
                sub = (
                    items.get(handling.substitute_item_id)
                    if handling.substitute_item_id
                    else None
                )
                if restriction.allergen_class:
                    sub_ok = (
                        sub is not None
                        and item_in_allergen_class(sub, restriction.allergen_class)
                        == "non_member"
                    )
                else:
                    sub_ok = sub is not None and sub.id != restriction.canonical_item_id
                if not sub_ok:
                    violations.append(
                        Violation(
                            code="invalid_substitute",
                            severity=err_severity,
                            entry_index=entry_index,
                            message=(
                                f"{base_msg}; proposed substitute is missing, unknown, "
                                f"or itself conflicting."
                            ),
                        )
                    )
                continue
            if handling.handling == "split":
                if not split_feasible(recipe, conflicting):
                    violations.append(
                        Violation(
                            code="split_not_feasible",
                            severity=err_severity,
                            entry_index=entry_index,
                            message=(
                                f"{base_msg}; split proposed but a conflicting ingredient "
                                f"enters at step 1 or an unknown step — no clean pull point."
                            ),
                        )
                    )
                continue
            # handling == "skip" resolves the conflict for that person.

    return violations
